src/main/scala/is/hail/utils/IntervalTree.scala

```diff
@@ -40,16 +49,45 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
       "includeStart" -> TBoolean().toJSON(includeStart),
       "includeEnd" -> TBoolean().toJSON(includeEnd))
 
-  def disjointAndLessThan(pord: ExtendedOrdering, other: Interval): Boolean = {
+  private def disjointAndLessThan(pord: ExtendedOrdering, other: Interval): Boolean = {
     val c = pord.compare(this.end, other.start)
     c < 0 || (c == 0 && (!this.includeEnd || !other.includeStart))
   }
 
-  def disjointAndGreaterThan(pord: ExtendedOrdering, other: Interval): Boolean = {
+  private def disjointAndGreaterThan(pord: ExtendedOrdering, other: Interval): Boolean = {
     val c = pord.compare(this.start, other.end)
     c > 0 || (c == 0 && (!this.includeStart || !other.includeEnd))
   }
 
+  private def adjacent(pord: ExtendedOrdering, other: Interval): Boolean = {
+      (pord.equiv(this.start, other.end) && (this.includeStart || other.includeEnd)) ||
+      (pord.equiv(this.end, other.start) && (this.includeEnd || other.includeStart))
+  }
+
+  def merge(pord: ExtendedOrdering, other: Interval): Option[Interval] = {
+    (definitelyEmpty(pord), other.definitelyEmpty(pord)) match {
+      case (true, true) => Some(this)
```

Comment:
Even if you keep the cases, it can be simplified using `(_, true) => Some(this)`. But this still feels unnatural to me, e.g. you're retesting definitelyEmpty inside mayOverlap. For the purposes of building the tree, do you really want the union of an empty interval and an interval to be the interval if they're not adjacent? Is it possible for the merge function to just be the last case:
```
if (mayOverlap(pord, other) || adjacent(pord, other)) {
          val min = Interval.ordering(pord, startPrimary = true).min(this, other).asInstanceOf[Interval]
          val max = Interval.ordering(pord, startPrimary = false).max(this, other).asInstanceOf[Interval]
          Some(Interval(min.start, max.end, min.includeStart, max.includeEnd))
        } else
          None
```

Reply:
It could be. For the purposes of building the tree, the empty cases don't matter because they get discarded. I like returning the non-empty interval because it sits closer to what I feel like `merge` or `union` *ought* to do, and the redundant emptiness checks aren't actually performing (possibly expensive) compares.